docs/design/2019-11-21-dynamic-configuration-change.md

```diff
@@ -0,0 +1,271 @@
+standard. For example, for the following configuration item, the name should be
+*readpool.storage.high-concurrency*.
+
+```toml
+[readpool.storage]
+high-concurrency = 4
+```
+
+There are two types for a configuration item: local and global. It is defined
+as follows:
+
+```proto
+message Local {
+    bytes component_id = 1;
+}
+
+message Global {
+    Component component = 1;
+}
+
+These two types are used to distinguish whether the configuration is shared by
+components. For example, the label configuration of TiKV is individual for each
+TiKV instance. So the type should be local. Each instance here is uniquely identified
+by the *component_id*, which can be obtained by hashing *IP: port*.
+Another example is the *raft-base-tick-interval* which is the configuration
+shared by all TiKV instances, so the type should be global. The definition of
+component type is as follows:
+
+```proto
+enum Component {
+    UNKNOWN = 0;
+    PD = 1;
+    TiKV = 2;
+    TiDB = 3;
+}
+```
+
+### Version Control
+
+To determine whether the corresponding component needs to update the
+configuration, the version field needs to be added to the request and response.
+The protocol definition is as follows:
+
+```proto
+message Version {
+    uint64 local = 1;
+    uint64 global = 2;
+}
+```
+
+For global configuration items, PD only maintains a set of global configuration
+items and a version variable, while for local, it needs to maintain a mapping
+from component ID to a set of local configuration items, and each set of local
+configuration items need to have an independent version variable.
+
+The following lists how version works In common scenarios:
+
+- New cluster. Both TiDB and TiKV use the default configuration and send it to
+PD to complete the registration. The registration needs to establish the mapping
+relationship between the component ID, version and local configuration. For
+customized requirements, such as modifying the size of block cache. It needs
+external tools to help modify them. After that, the global and local versions
+also need to be modified by adding one respectively according to the
+configuration option type. TiDB or TiKV obtains the configuration and version
+through *Get* method and decides whether to modify them by comparing with the
+local version. If the local version is smaller than the version carried by the
+response, the corresponding configuration options will be overwritten or merged
+and the version will be updated at the same time.
+
+- Add a new component or restart the component. The initialization of
+configuration calls the *Create* method. After receiving the request, PD first
+registers or queries the component ID. By comparing the version carried by the
+request with the version stored in PD, it determines whether to return the
+configuration of the component in the response. After receiving the reply, TiDB
+or TiKV decides whether to update the configuration or not after comparing with
```

Comment:
Suppose PD lost the configurations in a disaster, can operator make configuration changes after recovery? Seems like that the changes will be ignored by component because the version in PD is less than version in component.

Reply:
IIRC, the PD recover tool will recover the cluster-id and also pick a big enough alloc-id which needs bigger than the current allocated id. So I guess when storing configuration in PD, the corresponding config version needs to be handled the same way.